README.md
openstack-ifsc
==============

The problem to solve: buid a self-service IaaS for IFSC teachers (class activities) and students (remote labs), including template machines running open-source or proprietary software (with networked key).

What we have:
- 4 x HP Z220: 1 x Intel Xeon E3-1225v2, 4 x 4 GB DDR3-1600 nECC, 1 x SATA 1 TB @ 7.2K RPM (`sda`), 1 x SSD 240 GB (`sdb`), 2 x Gigabit Ethernet (Intel 82579LM, `em1`, and Realtek RTL8169, `p5p1`).
- 1 x router (VMWare, external).
- 2 x D-Link DGS-3100-24 (stacked).

What we're doing:
- 1 router (`roteador`) to control access to hypervisors and make things easier to VMs.
- 1 controller (`openstack0`), which runs the following services: SQL database, message queue, identity, image, object storage, block storage, dashboard, all service controllers (for compute and networking) and compute node (KVM).
- 3 compute nodes (`openstack1`, `openstack2` and `openstack3`) running compute (KVM) and network nodes.
- As each machine has locally 1 TB (and no RAID!), will be used 800 GB of it for Ceph (http://ceph.com) to gain availability (tested!) and, specially, a common backend for image, object and block storage without any SPOF. It will be 2 replicas of data in the cluster, so we're saying 4 disks x 800 GB / 2 replicas = 1600 GB. All SSD disks used are exclusive for VMs (root disks).

So... why not just use some ready deploy application, like ceph-ceploy, Juju or even puppetlabs-openstack?
- Well, because this is for learning purposes, and maybe one (of more) of these applications will be used in another place, but for now (we think) we need some pain to gain :-)

Installation
------------

Physical Mapping:
- `roteador`, `eth0`: virtual interface.
- `roteador`, `eth1`: virtual interface.
- `roteador`, `eth2`: virtual interface.
- `openstack0`, `em1`: switch stack, 2:01.
- `openstack0`, `p5p1`: switch stack, 2:22.
- `openstack1`, `em1`: switch stack, 2:10.
- `openstack1`, `p5p1`: switch stack, 2:13.
- `openstack2`, `em1`: switch stack, 2:19.
- `openstack2`, `p5p1`: switch stack, 2:21.
- `openstack3`, `em1`: switch stack, 2:04.
- `openstack3`, `p5p1`: switch stack, 2:11.

Bonding:
- `openstack0`: `em1` + `p5p1` = `bond0` (IEEE 802.1AX-2008, fast LACP).
- `openstack1`: `em1` + `p5p1` = `bond0` (IEEE 802.1AX-2008, fast LACP).
- `openstack2`: `em1` + `p5p1` = `bond0` (IEEE 802.1AX-2008, fast LACP).
- `openstack3`: `em1` + `p5p1` = `bond0` (IEEE 802.1AX-2008, fast LACP).

VLANs:
- 448:
  - Services: VMs' external network.
  - Addresses: 200.135.233.0/25.
    - `roteador`, eth2, untagged: ending with 126.
    - `openstack1`, `bond0`, tagged: no address.
    - `openstack2`, `bond0`, tagged: no address.
    - `openstack3`, `bond0`, tagged: no address.
- 449:
  - Services: hypervisors' remote control.
  - Addresses: 200.135.233.192/28.
    - `roteador`, eth1, untagged: 206.
	- `openstack0`, `bond0`, tagged: 200.
    - `openstack1`, `bond0`, tagged: 201.
    - `openstack2`, `bond0`, tagged: 202.
    - `openstack3`, `bond0`, tagged: 203.
- 450:
  - Services: control, centralized database, messages.
  - Addresses: 10.45.0.0/24, FC00:450::/64.
    - `openstack0`, `bond0`, tagged: 200.
    - `openstack1`, `bond0`, tagged: 201.
    - `openstack2`, `bond0`, tagged: 202.
    - `openstack3`, `bond0`, tagged: 203.
- 451:
  - Services: storage.
  - Addressess: 10.45.1.0/24, FC00:451::/64.
    - `openstack0`, `bond0`, tagged: 200.
    - `openstack1`, `bond0`, tagged: 201.
    - `openstack2`, `bond0`, tagged: 202.
    - `openstack3`, `bond0`, tagged: 203.
- 452:
  - Services: tunnel interface for networking service.
  - Addressess: 10.45.2.0/24, FC00:452::/64.
    - `openstack0`, `bond0`, tagged: 200.
    - `openstack1`, `bond0`, tagged: 201.
    - `openstack2`, `bond0`, tagged: 202.
    - `openstack3`, `bond0`, tagged: 203.

Operating System:
- Distribution: Ubuntu Server 14.04 LTS.
- Partitioning:
  - `sda` (1 TB):
    - Primary: `/`, ext4, 10 GB.
    - Primary: PV LVM, remaining disk space.
      - VG `openstack`.
        - LV `ceph`, xfs, 800 GB.
  - `sdb` (240 GB):
    - Primary: PV LVM, all disk space.
	  - VG `openstack-ssd`, which will be used for VMs' root disk.

Network and remote access: as the machines will stay out of physical access, the SSH server was installed with the operating system - DNS server was also installed in `openstack0`, in the beginning, to make things easier in the beginning. Some files were manually created to do so.

Initial files:
- `roteador`, `/etc/hostname`:
```
roteador
```

- `roteador`, `/etc/hosts`:
```
127.0.0.1	localhost
200.135.233.253	roteador.openstack.sj.ifsc.edu.br	roteador

# For Puppet
200.135.233.200 puppet

# For RabbitMQ
200.135.233.200 rabbitmq

# For nginx
# For nginx
200.135.233.200 nova-novncproxy.openstack.sj.ifsc.edu.br nova-novncproxy
200.135.233.200 radosgw.openstack.sj.ifsc.edu.br radosgw
200.135.233.203 dashboard.openstack.sj.ifsc.edu.br dashboard horizon horizon.openstack.sj.ifsc.edu.br openstack.sj.ifsc.edu.br
200.135.233.203 swift.openstack.sj.ifsc.edu.br swift
200.135.233.203 snmp-manager.openstack.sj.ifsc.edu.br snmp-manager zabbix

::1     localhost ip6-localhost ip6-loopback
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
```

- `roteador`, `/etc/network/interfaces`:
```
# Loopback
auto lo
iface lo inet loopback
iface lo inet6 loopback

# DMZ
auto eth0
iface eth0 inet static
	address 200.135.233.253
	netmask 255.255.255.252
	gateway 200.135.233.254
	dns-search openstack.sj.ifsc.edu.br
	dns-nameservers 200.135.37.72 200.135.37.65

# OpenStack-IFSC: real
auto eth1
iface eth1 inet static
	address 200.135.233.206
	netmask 255.255.255.240

# OpenStack-IFSC: VMs
auto eth2
iface eth2 inet static
	address 200.135.233.126
	netmask 255.255.255.128
```

- `openstack0`, `/etc/hostname`:
```
openstack0
```

- `openstack0`, `/etc/hosts`:
```
127.0.0.1 localhost openstack0.openstack.sj.ifsc.edu.br openstack0

::1     localhost ip6-localhost ip6-loopback openstack0.openstack.sj.ifsc.edu.br openstack0
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
```

- `openstack0`, `/etc/network/interfaces`:
```
# Loopback
auto lo
iface lo inet loopback
	dns-search openstack.sj.ifsc.edu.br
	dns-nameservers 10.45.0.201 10.45.0.202
iface lo inet6 loopback

# Onboard interface
auto em1 
iface em1 inet manual
	bond-master bond0
	bond-primary em1 p5p1
	bond-use-carrier 1

# Offboard interface
auto p5p1
iface p5p1 inet manual
	bond-master bond0
	bond-primary em1 p5p1
	bond-use-carrier 1

# Bonding
auto bond0
iface bond0 inet manual
	bond-slaves none
	bond-lacp-rate fast
	bond-mode 802.3ad
	bond-xmit_hash_policy layer2+3

# VMs
auto vlan448
iface vlan448 inet manual
	vlan-raw-device bond0

# Remote access
auto vlan449
iface vlan449 inet static
	vlan-raw-device bond0
	address 200.135.233.200
	netmask 255.255.255.240
	gateway 200.135.233.206

# Management
auto vlan450
iface vlan450 inet static
	vlan-raw-device bond0
	address 10.45.0.200
	netmask 255.255.255.0
iface vlan450 inet6 static
	address FC00:0450::200
	netmask 64
	accept_ra 0
	autconf 0
	scope site

# Storage
auto vlan451
iface vlan451 inet static
	vlan-raw-device bond0
	address 10.45.1.200
	netmask 255.255.255.0
iface vlan451 inet6 static
	address FC00:0451::200
	netmask 64
	accept_ra 0
	autconf 0
	scope site

# Tunnel interface
auto vlan452
iface vlan452 inet static
	vlan-raw-device bond0
	address 10.45.2.200
	netmask 255.255.255.0
iface vlan452 inet6 static
	address FC00:0452::200
	netmask 64
	accept_ra 0
	autconf 0
	scope site
```

- `openstack1`, `/etc/hostname`:
```
openstack1
```

- `openstack1`, `/etc/hosts`:
```
127.0.0.1 localhost openstack1.openstack.sj.ifsc.edu.br openstack1

::1     localhost ip6-localhost ip6-loopback openstack1.openstack.sj.ifsc.edu.br openstack1
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
```

- `openstack1`, `/etc/network/interfaces`:
```
# Loopback
auto lo
iface lo inet loopback
iface lo inet6 loopback

# Onboard interface
auto em1 
iface em1 inet manual
	bond-master bond0
	bond-primary em1 p5p1
	bond-use-carrier 1

# Offboard interface
auto p5p1
iface p5p1 inet manual
	bond-master bond0
	bond-primary em1 p5p1
	bond-use-carrier 1

# Bonding
auto bond0
iface bond0 inet manual
	bond-slaves none
	bond-lacp-rate fast
	bond-mode 802.3ad
	bond-xmit_hash_policy layer2+3

# VMs
auto vlan448
iface vlan448 inet manual
	vlan-raw-device bond0

# Remote access
auto vlan449
iface vlan449 inet static
	vlan-raw-device bond0
	address 200.135.233.201
	netmask 255.255.255.240
	gateway 200.135.233.206

# Management
auto vlan450
iface vlan450 inet static
	vlan-raw-device bond0
	address 10.45.0.201
	netmask 255.255.255.0
	dns-search openstack.sj.ifsc.edu.br
	dns-nameservers 10.45.0.201 10.45.0.202
iface vlan450 inet6 static
	address FC00:0450::201
	netmask 64
	accept_ra 0
	autconf 0
	scope site

# Storage
auto vlan451
iface vlan451 inet static
	vlan-raw-device bond0
	address 10.45.1.201
	netmask 255.255.255.0
iface vlan451 inet6 static
	address FC00:0451::201
	netmask 64
	accept_ra 0
	autconf 0
	scope site

# Tunnel interface
auto vlan452
iface vlan452 inet static
	vlan-raw-device bond0
	address 10.45.2.201
	netmask 255.255.255.0
iface vlan452 inet6 static
	address FC00:0452::201
	netmask 64
	accept_ra 0
	autconf 0
	scope site
```

- `openstack2`, `/etc/hostname`:
```
openstack2
```

- `openstack2`, `/etc/hosts`:
```
127.0.0.1 localhost openstack2.openstack.sj.ifsc.edu.br openstack2

::1     localhost ip6-localhost ip6-loopback openstack2.openstack.sj.ifsc.edu.br openstack2
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
```

- `openstack2`, `/etc/network/interfaces`:
```
# Loopback
auto lo
iface lo inet loopback
iface lo inet6 loopback

# Onboard interface
auto em1 
iface em1 inet manual
	bond-master bond0
	bond-primary em1 p5p1
	bond-use-carrier 1

# Offboard interface
auto p5p1
iface p5p1 inet manual
	bond-master bond0
	bond-primary em1 p5p1
	bond-use-carrier 1

# Bonding
auto bond0
iface bond0 inet manual
	bond-slaves none
	bond-lacp-rate fast
	bond-mode 802.3ad
	bond-xmit_hash_policy layer2+3

# VMs
auto vlan448
iface vlan448 inet manual
	vlan-raw-device bond0

# Remote access
auto vlan449
iface vlan449 inet static
	vlan-raw-device bond0
	address 200.135.233.202
	netmask 255.255.255.240
	gateway 200.135.233.206

# Management
auto vlan450
iface vlan450 inet static
	vlan-raw-device bond0
	address 10.45.0.202
	netmask 255.255.255.0
	dns-search openstack.sj.ifsc.edu.br
	dns-nameservers 10.45.0.201 10.45.0.202
iface vlan450 inet6 static
	address FC00:0450::202
	netmask 64
	accept_ra 0
	autconf 0
	scope site

# Storage
auto vlan451
iface vlan451 inet static
	vlan-raw-device bond0
	address 10.45.1.202
	netmask 255.255.255.0
iface vlan451 inet6 static
	address FC00:0451::202
	netmask 64
	accept_ra 0
	autconf 0
	scope site

# Tunnel interface
auto vlan452
iface vlan452 inet static
	vlan-raw-device bond0
	address 10.45.2.202
	netmask 255.255.255.0
iface vlan452 inet6 static
	address FC00:0452::202
	netmask 64
	accept_ra 0
	autconf 0
	scope site
```

- `openstack3`, `/etc/hostname`:
```
openstack3
```

- `openstack3`, `/etc/hosts`:
```
127.0.0.1 localhost openstack3.openstack.sj.ifsc.edu.br openstack3

::1     localhost ip6-localhost ip6-loopback openstack3.openstack.sj.ifsc.edu.br openstack3
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
```

- `openstack3`, `/etc/network/interfaces`:
```
# Loopback
auto lo
iface lo inet loopback
iface lo inet6 loopback

# Onboard interface
auto em1 
iface em1 inet manual
	bond-master bond0
	bond-primary em1 p5p1
	bond-use-carrier 1

# Offboard interface
auto p5p1
iface p5p1 inet manual
	bond-master bond0
	bond-primary em1 p5p1
	bond-use-carrier 1

# Bonding
auto bond0
iface bond0 inet manual
	bond-slaves none
	bond-lacp-rate fast
	bond-mode 802.3ad
	bond-xmit_hash_policy layer2+3

# VMs
auto vlan448
iface vlan448 inet manual
	vlan-raw-device bond0

# Remote access
auto vlan449
iface vlan449 inet static
	vlan-raw-device bond0
	address 200.135.233.203
	netmask 255.255.255.240
	gateway 200.135.233.206

# Management
auto vlan450
iface vlan450 inet static
	vlan-raw-device bond0
	address 10.45.0.203
	netmask 255.255.255.0
	dns-search openstack.sj.ifsc.edu.br
	dns-nameservers 10.45.0.201 10.45.0.202
iface vlan450 inet6 static
	address FC00:0450::203
	netmask 64
	accept_ra 0
	autconf 0
	scope site

# Storage
auto vlan451
iface vlan451 inet static
	vlan-raw-device bond0
	address 10.45.1.203
	netmask 255.255.255.0
iface vlan451 inet6 static
	address FC00:0451::203
	netmask 64
	accept_ra 0
	autconf 0
	scope site

# Tunnel interface
auto vlan452
iface vlan452 inet static
	vlan-raw-device bond0
	address 10.45.2.203
	netmask 255.255.255.0
iface vlan452 inet6 static
	address FC00:0452::203
	netmask 64
	accept_ra 0
	autconf 0
	scope site
```

And, to activate (repeated in all hosts):
```
ifdown -a
ifup -a
```
Note: two packages, `vlan` and `ifenslave`, are mandatory prior to put all interface up.

Uh-oh! Somehow, video card (nVidia) have some unknown issue when used with open-source drive [nouveau](http://nouveau.freedesktop.org/wiki/) (don't ask me why yet :-). Deactivated with file `/etc/modprobe.d/nvidia.conf`:
```
blacklist nvidia
```
and rebooted (I hate to do this, but it's safer to do).

# Puppet
Installed Puppet master in `openstack0`:
```
aptitude install puppetmaster
```
and agents:
```
aptitude install puppet
```

### Putting all together
- Put, in `roteador`, `openstack1`, `openstack2` and `openstack3` to wait for the master:
```
puppet agent --waitforcert 60
puppet agent --test
```
- In the master (`openstack0`):
```
puppet cert sign openstack1.sj.ifsc.edu.br
puppet cert sign openstack2.sj.ifsc.edu.br
puppet cert sign openstack3.sj.ifsc.edu.br
```
- And, finally, in every agent:
```
puppet agent --enable
puppet agent --test
```

## Puppet and git.
Ok, from this time master and agents can communicate. Now, it's time to combine Puppet and git (this project):
- In master:
```
cd /etc
git clone https://github.com/boidacarapreta/openstack-ifsc.git
mv puppet/* openstack-ifsc/
ln -s openstack-ifsc/puppet puppet
```
Because there are many editors, including me (@boidacarapreta), some file have been modified in its modes:
```
chgrp -R git /etc/openstack-ifsc/
chmod -R g+w /etc/openstack-ifsc/
```

## Internal services
NTP, SNMP and even SSH (reconfigured) was installed and configured using puppet. Check the code :-)


HISTORY.md
Icehouse
===
The first release.

v1.0
===
Environment tagged after a half-year with no critical bugs.

From Icehouse to Juno
===
Upgrade documented in commit 39bd39e85cc8cba1bc65e73fd33f89f20b6aff4a.

Some databases need to be upgraded:
```
su -s /bin/bash - keystone -c "/usr/bin/keystone-manage db_sync"
su -s /bin/bash - glance -c "/usr/bin/glance-manage db_sync"
su -s /bin/bash - cinder -c "/usr/bin/cinder-manage db sync"
su -s /bin/bash - nova -c "/usr/bin/nova-manage db sync"
su -s /bin/bash - heat -c "/usr/bin/heat-manage db_sync"
su -s /bin/bash - ceilometer -c "/usr/bin/ceilometer-dbsync"
su -s /bin/bash - neutron -c "/usr/bin/neutron-db-manage --config-file /etc/neutron/neutron.conf --config-file /etc/neutron/plugins/ml2/ml2_conf.ini --config-file /etc/neutron/lbaas_agent.ini --config-file /etc/neutron/vpn_agent.ini --config-file /etc/neutron/l3_agent.ini --config-file /etc/neutron/fwaas_driver.ini stamp icehouse"
su -s /bin/bash - neutron -c "/usr/bin/neutron-db-manage --config-file /etc/neutron/neutron.conf --config-file /etc/neutron/plugins/ml2/ml2_conf.ini --config-file /etc/neutron/lbaas_agent.ini --config-file /etc/neutron/vpn_agent.ini --config-file /etc/neutron/l3_agent.ini --config-file /etc/neutron/fwaas_driver.ini upgrade head"
```

v1.1
===
Environment tagged after a half-year with no critical bugs.

From Juno to Kilo
===
Commits 0a9de2d65bf6d491818847611fdddb8055f8cfd6 to ab141bdef99d008750dc76c120e8c3e599337062 (and counting).

Some databases need to be upgraded, as the former version upgrade:
```
su -s /bin/bash - keystone -c "/usr/bin/keystone-manage db_sync"
su -s /bin/bash - glance -c "/usr/bin/glance-manage db_sync"
su -s /bin/bash - cinder -c "/usr/bin/cinder-manage db sync"
su -s /bin/bash - nova -c "/usr/bin/nova-manage db sync"
su -s /bin/bash - heat -c "/usr/bin/heat-manage db_sync"
su -s /bin/bash - ceilometer -c "/usr/bin/ceilometer-dbsync"
```
In Neutron, I had a lot of errors. I forgot to stamp the database, I guess. So, I had to fix manually (Python is my friend: many, many files in `/usr/lib/python2.7/dist-packages/neutron/db/migration/alembic_migrations/versions`). After it, I could run the regular commands:
```
su -s /bin/bash - neutron -c "/usr/bin/neutron-db-manage --config-file /etc/neutron/neutron.conf --config-file /etc/neutron/plugins/ml2/ml2_conf.ini --config-file /etc/neutron/lbaas_agent.ini --config-file /etc/neutron/vpn_agent.ini --config-file /etc/neutron/l3_agent.ini --config-file /etc/neutron/fwaas_driver.ini stamp juno"
su -s /bin/bash - neutron -c "/usr/bin/neutron-db-manage --config-file /etc/neutron/neutron.conf --config-file /etc/neutron/plugins/ml2/ml2_conf.ini --config-file /etc/neutron/lbaas_agent.ini --config-file /etc/neutron/vpn_agent.ini --config-file /etc/neutron/l3_agent.ini --config-file /etc/neutron/fwaas_driver.ini upgrade head"
```

VPNaaS, a Neutron plugin, was introduced in bdfe6b847e2bd3f99ae2cad12a1ff0199861fa7c, but no one really used it. So, I deactivated it to a simpler Neutron configuration.


puppet/modules/openstack-horizon/files/local_settings.py
import os

from django.utils.translation import ugettext_lazy as _
from openstack_dashboard import exceptions

DEBUG = False
TEMPLATE_DEBUG = DEBUG

# https://wiki.openstack.org/wiki/ReleaseNotes/Kilo#Key_New_Features_4
LAUNCH_INSTANCE_NG_ENABLED = True

HORIZON_CONFIG = {
	'dashboards': ('project', 'admin', 'settings',),
	'default_dashboard': 'project',
	'user_home': 'openstack_dashboard.views.get_user_home',
	'ajax_queue_limit': 10,
	'auto_fade_alerts': {
		'delay': 3000,
		'fade_duration': 1500,
		'types': ['alert-success', 'alert-info']
	},
	'help_url': "http://docs.openstack.org",
	'exceptions': {
        'recoverable': exceptions.RECOVERABLE,
        'not_found': exceptions.NOT_FOUND,
        'unauthorized': exceptions.UNAUTHORIZED
    },
}

LOCAL_PATH = os.path.dirname(os.path.abspath(__file__))

from horizon.utils import secret_key
SECRET_KEY = secret_key.generate_or_read_from_file('/etc/openstack-dashboard/secret_key')

SESSION_ENGINE = 'django.contrib.sessions.backends.cache'
CACHES = {
   'default': {
	   'BACKEND' : 'django.core.cache.backends.memcached.MemcachedCache',
	   'LOCATION' : 'memcached.openstack.sj.ifsc.edu.br:11211',
   }
}

try:
	from ubuntu_theme import *
except ImportError:
	pass

LOGIN_URL='/horizon/auth/login/'
LOGOUT_URL='/horizon/auth/logout/'
LOGIN_REDIRECT_URL='https://openstack.sj.ifsc.edu.br/horizon/'
COMPRESS_OFFLINE = True
ALLOWED_HOSTS = '*'
EMAIL_BACKEND = 'django.core.mail.backends.console.EmailBackend'
OPENSTACK_HOST = "keystone.openstack.sj.ifsc.edu.br"
OPENSTACK_KEYSTONE_DEFAULT_ROLE = "_member_"

# Based on https://docs.hpcloud.com/helion/openstack/1.1/services/identity/configure/
OPENSTACK_API_VERSIONS = { "identity": 3, }
OPENSTACK_KEYSTONE_URL = "http://%s:5000/v3" % OPENSTACK_HOST

OPENSTACK_KEYSTONE_BACKEND = {
	'name': 'native',
	'can_edit_user': True,
	'can_edit_group': True,
	'can_edit_project': True,
	'can_edit_domain': True,
	'can_edit_role': True
}

OPENSTACK_HYPERVISOR_FEATURES = {
	'can_set_mount_point': False,
	'can_set_password': False,
}

OPENSTACK_CINDER_FEATURES = {
	'enable_backup': True,
}

OPENSTACK_NEUTRON_NETWORK = {
	'enable_lb': True,
	'enable_firewall': False,
	'enable_quotas': True,
	'enable_router': True,
	'enable_ipv6': True,
	'enable_distributed_router': False,
	'enable_ha_router': False,
	'enable_vpn': False,
	'profile_support': None,
	'supported_provider_types': ['*'],
}

IMAGE_CUSTOM_PROPERTY_TITLES = {
	"architecture": _("Architecture"),
	"kernel_id": _("Kernel ID"),
	"ramdisk_id": _("Ramdisk ID"),
	"image_state": _("Euca2ools state"),
	"project_id": _("Project ID"),
	"image_type": _("Image Type")
}

API_RESULT_LIMIT = 1000
API_RESULT_PAGE_SIZE = 20
TIME_ZONE = "UTC"

LOGGING = {
	'version': 1,
	'disable_existing_loggers': False,
	'handlers': {
		'null': {
			'level': 'DEBUG',
			'class': 'django.utils.log.NullHandler',
		},
		'console': {
			'level': 'INFO',
			'class': 'logging.StreamHandler',
		},
	},
	'loggers': {
		'django.db.backends': {
			'handlers': ['null'],
			'propagate': False,
		},
		'requests': {
			'handlers': ['null'],
			'propagate': False,
		},
		'horizon': {
			'handlers': ['console'],
			'level': 'DEBUG',
			'propagate': False,
		},
		'openstack_dashboard': {
			'handlers': ['console'],
			'level': 'DEBUG',
			'propagate': False,
		},
		'novaclient': {
			'handlers': ['console'],
			'level': 'DEBUG',
			'propagate': False,
		},
		'cinderclient': {
			'handlers': ['console'],
			'level': 'DEBUG',
			'propagate': False,
		},
		'keystoneclient': {
			'handlers': ['console'],
			'level': 'DEBUG',
			'propagate': False,
		},
		'glanceclient': {
			'handlers': ['console'],
			'level': 'DEBUG',
			'propagate': False,
		},
		'neutronclient': {
			'handlers': ['console'],
			'level': 'DEBUG',
			'propagate': False,
		},
		'heatclient': {
			'handlers': ['console'],
			'level': 'DEBUG',
			'propagate': False,
		},
		'ceilometerclient': {
			'handlers': ['console'],
			'level': 'DEBUG',
			'propagate': False,
		},
		'troveclient': {
			'handlers': ['console'],
			'level': 'DEBUG',
			'propagate': False,
		},
		'swiftclient': {
			'handlers': ['console'],
			'level': 'DEBUG',
			'propagate': False,
		},
		'openstack_auth': {
			'handlers': ['console'],
			'level': 'DEBUG',
			'propagate': False,
		},
		'nose.plugins.manager': {
			'handlers': ['console'],
			'level': 'DEBUG',
			'propagate': False,
		},
		'django': {
			'handlers': ['console'],
			'level': 'DEBUG',
			'propagate': False,
		},
		'iso8601': {
			'handlers': ['null'],
			'propagate': False,
		},
		'scss': {
			'handlers': ['null'],
			'propagate': False,
		},
	}
}

SECURITY_GROUP_RULES = {
	'all_tcp': {
		'name': _('All TCP'),
		'ip_protocol': 'tcp',
		'from_port': '1',
		'to_port': '65535',
	},
	'all_udp': {
		'name': _('All UDP'),
		'ip_protocol': 'udp',
		'from_port': '1',
		'to_port': '65535',
	},
	'all_icmp': {
		'name': _('All ICMP'),
		'ip_protocol': 'icmp',
		'from_port': '-1',
		'to_port': '-1',
	},
	'ssh': {
		'name': 'SSH',
		'ip_protocol': 'tcp',
		'from_port': '22',
		'to_port': '22',
	},
	'smtp': {
		'name': 'SMTP',
		'ip_protocol': 'tcp',
		'from_port': '25',
		'to_port': '25',
	},
	'dns': {
		'name': 'DNS',
		'ip_protocol': 'tcp',
		'from_port': '53',
		'to_port': '53',
	},
	'http': {
		'name': 'HTTP',
		'ip_protocol': 'tcp',
		'from_port': '80',
		'to_port': '80',
	},
	'pop3': {
		'name': 'POP3',
		'ip_protocol': 'tcp',
		'from_port': '110',
		'to_port': '110',
	},
	'imap': {
		'name': 'IMAP',
		'ip_protocol': 'tcp',
		'from_port': '143',
		'to_port': '143',
	},
	'ldap': {
		'name': 'LDAP',
		'ip_protocol': 'tcp',
		'from_port': '389',
		'to_port': '389',
	},
	'https': {
		'name': 'HTTPS',
		'ip_protocol': 'tcp',
		'from_port': '443',
		'to_port': '443',
	},
	'smtps': {
		'name': 'SMTPS',
		'ip_protocol': 'tcp',
		'from_port': '465',
		'to_port': '465',
	},
	'imaps': {
		'name': 'IMAPS',
		'ip_protocol': 'tcp',
		'from_port': '993',
		'to_port': '993',
	},
	'pop3s': {
		'name': 'POP3S',
		'ip_protocol': 'tcp',
		'from_port': '995',
		'to_port': '995',
	},
	'ms_sql': {
		'name': 'MS SQL',
		'ip_protocol': 'tcp',
		'from_port': '1433',
		'to_port': '1433',
	},
	'mysql': {
		'name': 'MYSQL',
		'ip_protocol': 'tcp',
		'from_port': '3306',
		'to_port': '3306',
	},
	'rdp': {
		'name': 'RDP',
		'ip_protocol': 'tcp',
		'from_port': '3389',
		'to_port': '3389',
	},
}

FLAVOR_EXTRA_KEYS = {
	'flavor_keys': [
		('quota:read_bytes_sec', _('Quota: Read bytes')),
		('quota:write_bytes_sec', _('Quota: Write bytes')),
		('quota:cpu_quota', _('Quota: CPU')),
		('quota:cpu_period', _('Quota: CPU period')),
		('quota:inbound_average', _('Quota: Inbound average')),
		('quota:outbound_average', _('Quota: Outbound average')),
	]
}
